--- a/jp/ngt/rtm/modelpack/cfg/ResourceConfig.java
+++ b/jp/ngt/rtm/modelpack/cfg/ResourceConfig.java
@@ -5,14 +5,17 @@
 public abstract class ResourceConfig {
    public short version;
    public boolean useCustomColor;
    public String tags;
    public ResourceConfig.DMInitValue[] defaultValues;
+   public java.io.File file;
    /** @deprecated */
    @Deprecated
    public String defaultData;
 
+   public boolean synchronousLoading;
+
    public abstract String getName();
 
    public void init() {
       if (this.defaultValues == null && this.defaultData != null) {
          String[][] astring = DataMap.convertArg(this.defaultData);
@@ -28,14 +31,18 @@
       }
 
    }
 
    protected String fixSoundPath(String path) {
+      return fixSoundPath(path, null);
+   }
+
+   protected String fixSoundPath(String path, String defaults) {
       if (path != null && path.length() != 0) {
          return !path.contains(":") ? "rtm:" + path : path;
       } else {
-         return null;
+         return defaults;
       }
    }
 
    public class DMInitValue {
       public String type;
