--- a/jp/ngt/rtm/modelpack/init/MPLFrame.java
+++ b/jp/ngt/rtm/modelpack/init/MPLFrame.java
@@ -1,19 +1,23 @@
 package jp.ngt.rtm.modelpack.init;
 
 import javax.swing.JFrame;
+
+import net.minecraft.client.Minecraft;
 import org.lwjgl.opengl.Display;
 
 public final class MPLFrame extends JFrame {
    public static final int FRAME_WIDTH = 480;
    public static final int FRAME_HEIGHT = 360;
 
    public MPLFrame(int scale) {
       super("RealTrainMod");
+
+      Minecraft minecraft = Minecraft.getMinecraft();
+      int windowPosX = Display.getX() + (minecraft.displayWidth / 2) - (FRAME_WIDTH / 2);
+      int windowPosY = Display.getY() + (minecraft.displayHeight / 2) - (FRAME_HEIGHT / 2);
+
       this.setDefaultCloseOperation(0);
-      this.setSize(480 * scale, 360 * scale);
-      int i = Display.getX();
-      int j = Display.getY();
-      this.setLocation(i, j);
-      this.setResizable(false);
+      this.setBounds(windowPosX, windowPosY, FRAME_WIDTH * scale, FRAME_HEIGHT * scale);
+      this.setResizable(true);
    }
 }
