--- a/jp/ngt/rtm/entity/train/util/EnumNotch.java
+++ b/jp/ngt/rtm/entity/train/util/EnumNotch.java
@@ -86,18 +86,9 @@
 
          return inertia;
       }
    }
 
-   public static float getAcceleration(int notch, float prevSpeed, TrainConfig cfg) {
-      if (notch == 0) {
-         return 0.0F;
-      } else if (notch > 0) {
-         --notch;
-         return prevSpeed >= cfg.maxSpeed[notch] ? 0.0F : cfg.accelerateion;
-      } else {
-         float f = getNotch(notch).acceleration;
-         float f1 = Math.abs(prevSpeed);
-         return f1 + f < 0.0F ? -f1 : f;
-      }
+   public static float getAcceleration(int notch, float prevSpeed, TrainConfig cfg, jp.ngt.rtm.entity.train.EntityTrainBase train) {
+      return com.anatawa12.fixRtm.rtm.entity.train.util.EnumNotchKt.getAcceleration(notch, prevSpeed, cfg, train);
    }
 }
