--- a/jp/ngt/rtm/entity/train/parts/EntityFloor.java
+++ b/jp/ngt/rtm/entity/train/parts/EntityFloor.java
@@ -1,8 +1,9 @@
 package jp.ngt.rtm.entity.train.parts;
 
 import jp.ngt.ngtlib.util.NGTUtil;
+import jp.ngt.rtm.entity.train.EntityTrainBase;
 import jp.ngt.rtm.entity.vehicle.EntityVehicleBase;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.Items;
@@ -11,10 +12,11 @@
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.EnumHand;
 import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class EntityFloor extends EntityVehiclePart {
@@ -120,6 +122,26 @@
    @SideOnly(Side.CLIENT)
    public int getBrightnessForRender() {
       EntityVehicleBase entityvehiclebase = this.getVehicle();
       return entityvehiclebase != null ? entityvehiclebase.getBrightnessForRender() : super.getBrightnessForRender();
    }
+
+   @Override
+   public void updatePassenger(Entity passenger) {
+      super.updatePassenger(passenger);
+
+      if (!this.isPassenger(passenger)) return;
+
+      EntityVehicleBase<?> vehicle = this.getVehicle();
+
+      if (!(passenger instanceof EntityPlayer) || !(vehicle instanceof EntityTrainBase)) return;
+
+      EntityPlayer player = ((EntityPlayer) passenger);
+
+      float yaw = MathHelper.wrapDegrees(vehicle.rotationYaw - vehicle.prevRotationYaw);
+      float pitch = MathHelper.wrapDegrees(vehicle.rotationPitch - vehicle.prevRotationPitch);
+
+      player.renderYawOffset -= yaw;
+      player.rotationYaw -= yaw;
+      player.rotationPitch -= pitch;
+   }
 }
