--- a/jp/ngt/rtm/gui/GuiIngameCustom.java
+++ b/jp/ngt/rtm/gui/GuiIngameCustom.java
@@ -98,10 +98,13 @@
       FontRenderer fontrenderer = this.mc.fontRenderer;
       fontrenderer.drawStringWithShadow(String.valueOf(this.getSpeed(train)), (float)(i - 138), (float)(this.height - 11), 65280);
       fontrenderer.drawStringWithShadow(String.valueOf(this.getBrake(train)), (float)(i - 178), (float)(this.height - 11), 65280);
       fontrenderer.drawStringWithShadow(String.valueOf(this.getWorldTime()), (float)(i + 130), (float)(this.height - 40), 65280);
       fontrenderer.drawStringWithShadow(this.getTime(), (float)(i + 130), (float)(this.height - 30), 65280);
+
+      int doorLampColor = train.doorMoveL + train.doorMoveR == 0 ? 0x00ffff : 0x505050;
+      fontrenderer.drawStringWithShadow("Door", (float)(i + 130), (float)(this.height - 20), doorLampColor);
    }
 
    private void renderTimeTable(EntityTrainBase train) {
       String s = train.getResourceState().getName();
       if (this.timeTable == null || !s.equals(this.timeTable.train)) {
@@ -191,12 +194,11 @@
 
    public String getTime() {
       int i = this.getWorldTime();
       int j = (i / 1000 + 6) % 24;
       int k = (int)((float)(i % 1000) * 0.06F);
-      StringBuilder stringbuilder = (new StringBuilder(String.valueOf(j))).append(":").append(k);
-      return stringbuilder.toString();
+      return String.format("%02d:%02d", j, k);
    }
 
    private void setScale(ScaledResolution par1) {
       this.width = par1.getScaledWidth();
       this.height = par1.getScaledHeight();
@@ -210,11 +212,11 @@
       float f1 = f * (float)train.brakeAirCount / 2880.0F;
       this.drawMeter(j, k, 32, 32, 48, f1, 512);
       float f2 = f * (float)this.getBrake(train) / 432.0F;
       this.drawMeter(j, k, 32, 0, 48, f2, 512);
       float f3 = (train.getResourceState().getResourceSet().getConfig()).maxSpeed[4];
-      float f4 = 270.0F * train.getSpeed() / f3;
+      float f4 = 270.0F * Math.abs(train.getSpeed()) / f3;
       this.drawMeter(i - 136, this.height - 29, 32, 64, 48, f4, 512);
       this.drawLever(train);
    }
 
    private void drawLever(EntityTrainBase train) {
