--- a/jp/ngt/rtm/electric/BlockSpeaker.java
+++ b/jp/ngt/rtm/electric/BlockSpeaker.java
@@ -5,14 +5,16 @@
 import jp.ngt.rtm.RTMItem;
 import jp.ngt.rtm.block.BlockMachineBase;
 import jp.ngt.rtm.item.ItemInstalledObject;
 import net.minecraft.block.SoundType;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.World;
 
 public class BlockSpeaker extends BlockMachineBase implements IBlockConnective {
    public BlockSpeaker() {
       super(Material.GLASS);
@@ -45,6 +47,11 @@
    }
 
    public boolean canConnect(World world, int x, int y, int z) {
       return true;
    }
+
+   @Override
+   public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World world, BlockPos pos, EntityPlayer player) {
+      return com.anatawa12.fixRtm.rtm.block.BlockOrnamentMain.getPickBlock(world, pos, ItemInstalledObject.IstlObjType.SPEAKER);
+   }
 }
