--- a/jp/ngt/rtm/block/BlockDecoration.java
+++ b/jp/ngt/rtm/block/BlockDecoration.java
@@ -1,10 +1,12 @@
 package jp.ngt.rtm.block;
 
 import jp.ngt.ngtlib.block.BlockContainerCustom;
+import jp.ngt.rtm.RTMItem;
 import jp.ngt.rtm.block.tileentity.TileEntityDecoration;
 import net.minecraft.block.material.Material;
+import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.world.World;
 
 public class BlockDecoration extends BlockContainerCustom {
    public BlockDecoration() {
@@ -12,6 +14,11 @@
    }
 
    public TileEntity createNewTileEntity(World world, int meta) {
       return new TileEntityDecoration();
    }
+
+   @Override
+   protected ItemStack getItem(int damage) {
+      return new ItemStack(RTMItem.decoration_block);
+   }
 }
