--- a/jp/ngt/rtm/block/BlockCrossingGate.java
+++ b/jp/ngt/rtm/block/BlockCrossingGate.java
@@ -6,14 +6,16 @@
 import jp.ngt.rtm.block.tileentity.TileEntityCrossingGate;
 import jp.ngt.rtm.item.ItemInstalledObject;
 import net.minecraft.block.SoundType;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.World;
 
 public class BlockCrossingGate extends BlockMachineBase {
    public BlockCrossingGate() {
       super(Material.ROCK);
@@ -46,6 +48,11 @@
 
    private void checkPower(World world, int x, int y, int z) {
       TileEntityCrossingGate tileentitycrossinggate = (TileEntityCrossingGate)BlockUtil.getTileEntity(world, x, y, z);
       tileentitycrossinggate.isGettingPower = world.getRedstonePowerFromNeighbors(new BlockPos(x, y, z)) > 0;
    }
+
+   @Override
+   public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World world, BlockPos pos, EntityPlayer player) {
+      return com.anatawa12.fixRtm.rtm.block.BlockOrnamentMain.getPickBlock(world, pos, ItemInstalledObject.IstlObjType.CROSSING);
+   }
 }
