--- a/jp/ngt/rtm/block/BlockPoint.java
+++ b/jp/ngt/rtm/block/BlockPoint.java
@@ -3,16 +3,19 @@
 import jp.ngt.ngtlib.block.BlockArgHolder;
 import jp.ngt.ngtlib.block.BlockUtil;
 import jp.ngt.ngtlib.util.NGTUtil;
 import jp.ngt.rtm.RTMItem;
 import jp.ngt.rtm.block.tileentity.TileEntityPoint;
+import jp.ngt.rtm.item.ItemInstalledObject;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.World;
 
 public class BlockPoint extends BlockMachineBase {
    public BlockPoint() {
       super(Material.ROCK);
@@ -65,6 +68,11 @@
    }
 
    public boolean canProvidePower(IBlockState state) {
       return true;
    }
+
+   @Override
+   public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World world, BlockPos pos, EntityPlayer player) {
+      return com.anatawa12.fixRtm.rtm.block.BlockOrnamentMain.getPickBlock(world, pos, ItemInstalledObject.IstlObjType.POINT);
+   }
 }
