--- a/jp/ngt/rtm/rail/TileEntityMarker.java
+++ b/jp/ngt/rtm/rail/TileEntityMarker.java
@@ -1,9 +1,11 @@
 package jp.ngt.rtm.rail;
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.UUID;
+
 import jp.ngt.ngtlib.block.BlockUtil;
 import jp.ngt.ngtlib.block.TileEntityCustom;
 import jp.ngt.ngtlib.io.NGTLog;
 import jp.ngt.rtm.RTMBlock;
 import jp.ngt.rtm.gui.InternalButton;
@@ -14,16 +16,19 @@
 import jp.ngt.rtm.rail.util.RailMap;
 import jp.ngt.rtm.rail.util.RailMapBasic;
 import jp.ngt.rtm.rail.util.RailMapCustom;
 import jp.ngt.rtm.rail.util.RailPosition;
 import jp.ngt.rtm.rail.util.SwitchType;
+import net.minecraft.client.Minecraft;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.ITickable;
 import net.minecraft.util.math.BlockPos;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public class TileEntityMarker extends TileEntityCustom implements ITickable {
    private static final int SEARCH_COUNT = 40;
    private RailPosition rp;
    private BlockPos startPos;
@@ -40,45 +45,61 @@
    @SideOnly(Side.CLIENT)
    public InternalButton[] buttons;
    @SideOnly(Side.CLIENT)
    public float[][][] linePos;
    private int count;
+   private boolean isFirstTime = true;
+   @Nullable
+   private UUID owner;
 
    public TileEntityMarker() {
       this.markerState = MarkerState.DISTANCE.set(this.markerState, true);
       this.markerState = MarkerState.GRID.set(this.markerState, false);
       this.markerState = MarkerState.LINE1.set(this.markerState, false);
       this.markerState = MarkerState.LINE2.set(this.markerState, false);
       this.markerState = MarkerState.ANCHOR21.set(this.markerState, false);
+      this.markerState = MarkerState.FIT_NEIGHBOR.set(this.markerState, true);
    }
 
    public void readFromNBT(NBTTagCompound nbt) {
       super.readFromNBT(nbt);
       if (nbt.hasKey("RP")) {
-         this.rp = RailPosition.readFromNBT(nbt.getCompoundTag("RP"));
+         this.rp = RailPosition.readFromNBT(nbt.getCompoundTag("RP"), this.rp);
       }
 
+      if (nbt.hasUniqueId("Owner")) {
+         this.owner = nbt.getUniqueId("Owner");
+      }
    }
 
    public NBTTagCompound writeToNBT(NBTTagCompound nbt) {
       super.writeToNBT(nbt);
       if (this.rp != null) {
          nbt.setTag("RP", this.rp.writeToNBT());
       }
 
+      if (this.owner != null) {
+         nbt.setUniqueId("Owner", this.owner);
+      }
+
       return nbt;
    }
 
    public void update() {
-      if (this.rp == null) {
+      if (this.rp == null && this.owner != null) {
          byte b0 = BlockMarker.getMarkerDir(this.getBlockType(), this.getBlockMetadata());
          byte b1 = (byte)(this.getBlockType() == RTMBlock.markerSwitch ? 1 : 0);
          this.rp = new RailPosition(this.getPos().getX(), this.getPos().getY(), this.getPos().getZ(), b0, b1);
-         if (this.getWorld().isRemote) {
-            for(TileEntity tileentity : this.getWorld().loadedTileEntityList) {
-               if (tileentity instanceof TileEntityMarker) {
-                  ((TileEntityMarker)tileentity).searchOtherMarkers();
+      }
+
+      if (isFirstTime) {
+         isFirstTime = false;
+
+         if (world.isRemote) {
+            for (TileEntity tileEntity : world.loadedTileEntityList) {
+               if (tileEntity instanceof TileEntityMarker) {
+                  ((TileEntityMarker) tileEntity).searchOtherMarkers();
                }
             }
          }
       }
 
@@ -88,15 +109,16 @@
             this.count = 0;
          }
 
          ++this.count;
       }
-
    }
 
+   @SideOnly(Side.CLIENT)
    private void searchOtherMarkers() {
-      ((BlockMarker)this.getBlockType()).makeRailMap(this, this.getX(), this.getY(), this.getZ());
+      UUID uuid = Minecraft.getMinecraft().player.getUniqueID();
+      ((BlockMarker)this.getBlockType()).makeRailMap(this, uuid, this.getX(), this.getY(), this.getZ());
    }
 
    private void updateStartPos() {
       if (this.startPos != null) {
          TileEntity tileentity = this.getWorld().getTileEntity(this.startPos);
@@ -139,11 +161,11 @@
 
          for(int i = 0; i < arailmap.length; ++i) {
             RailPosition railposition = this.railMaps[i].getStartRP();
             RailPosition railposition1 = this.railMaps[i].getEndRP();
             railposition1.cantCenter = -railposition.cantCenter;
-            arailmap[i] = new RailMapBasic(railposition, railposition1);
+            arailmap[i] = new RailMapBasic(railposition, railposition1, RailMapBasic.fixRTMRailMapVersionCurrent);
          }
 
          this.railMaps = arailmap;
          this.linePos = (float[][][])null;
          this.createGrids();
@@ -171,26 +193,26 @@
       } else if (list.size() == 2) {
          if (list.get(0) != null && list.get(1) != null) {
             RailPosition railposition2 = this.getMarkerRP((BlockPos)list.get(0));
             RailPosition railposition3 = this.getMarkerRP((BlockPos)list.get(1));
             if (railposition2 != null && railposition3 != null) {
-               RailMap railmap1 = new RailMapBasic(railposition2, railposition3);
+               RailMap railmap1 = new RailMapBasic(railposition2, railposition3, RailMapBasic.fixRTMRailMapVersionCurrent);
                this.railMaps = new RailMap[]{railmap1};
                blockpos = new BlockPos(railposition2.blockX, railposition2.blockY, railposition2.blockZ);
             }
          }
       } else {
-         List<RailPosition> list = new ArrayList<>();
+         List<RailPosition> list3 = new ArrayList<>();
 
          for(BlockPos blockpos2 : list) {
             RailPosition railposition1 = this.getMarkerRP(blockpos2);
             if (railposition1 != null) {
-               list.add(railposition1);
+               list3.add(railposition1);
             }
          }
 
-         SwitchType switchtype = (new RailMaker(this.getWorld(), list)).getSwitch();
+         SwitchType switchtype = (new RailMaker(this.getWorld(), list3, RailMapBasic.fixRTMRailMapVersionCurrent)).getSwitch();
          if (switchtype != null) {
             this.railMaps = switchtype.getAllRailMap();
             if (this.railMaps != null) {
                RailPosition railposition4 = this.railMaps[0].getStartRP();
                blockpos = new BlockPos(railposition4.blockX, railposition4.blockY, railposition4.blockZ);
@@ -223,11 +245,11 @@
       }
 
    }
 
    private void setStartPos(BlockPos pos, RailMap[] maps) {
-      NGTLog.debug("[Marker] Start pos %s", new Object[]{pos.toString()});
+      NGTLog.debug("[Marker] Start pos %s for %s", pos.toString(), this.getPos());
       this.startPos = pos;
       this.railMaps = maps;
       if (!this.isCoreMarker()) {
          this.markerPosList.clear();
          this.grid = null;
@@ -302,6 +324,15 @@
 
    public String getStateString(MarkerState state) {
       boolean flag = state.get(this.markerState);
       return String.format("%s : %s", state.toString(), flag ? "ON" : "OFF");
    }
+
+   @Nullable
+   public UUID getOwner() {
+      return owner;
+   }
+
+   public void setOwner(@NotNull UUID uuid) {
+      owner = uuid;
+   }
 }
